crates/subspace-block-relay/src/utils.rs

```diff
@@ -0,0 +1,158 @@
+//! Common utils.
+
+use crate::NetworkStub;
+use async_trait::async_trait;
+use codec::{Decode, Encode};
+use futures::channel::oneshot::{self, Canceled};
+use parking_lot::Mutex;
+use sc_network::request_responses::IfDisconnected;
+use sc_network::types::ProtocolName;
+use sc_network::{NetworkRequest, NetworkService, OutboundFailure, PeerId, RequestFailure};
+use sp_runtime::traits::Block as BlockT;
+use std::sync::Arc;
+
+type NetworkHandle<Block> = Arc<NetworkService<Block, <Block as BlockT>::Hash>>;
+
+/// The message sent to the server as part of the request/response.
+#[derive(Encode, Decode)]
+pub(crate) struct ServerMessage {
+    /// The serialized messages
+    pub(crate) message: Vec<u8>,
+
+    /// If the message is meant for the protocol component on the server side.
+    pub(crate) is_protocol_message: bool,
+}
+
+/// Wrapper to work around the circular dependency in substrate:
+/// `build_network()` requires the block relay to be passed in,
+/// which internally needs the network handle. `set()` is
+/// used to fill in the network after the network is created.
+pub struct NetworkWrapper<Block: BlockT> {
+    network: Mutex<Option<NetworkHandle<Block>>>,
+}
```

Comment:
yes this is terrible right now.  substrate handles this for the default block handler by using https://paritytech.github.io/substrate/master/sc_network_sync/service/network/struct.NetworkServiceHandle.html, which is just a wrapper around the send handle of a channel. This happens inside `build_network()`, but does not work for custom handlers like ours.

My initial patch to upstream required `download_block()` to take network handle for same reason, but the upstream team did not like it. 

I had considered different options, but decided against it as these would need more changes on upstream side: 
1. We could push this "inject network handle after creation" to upstream, but would end up looking similar. 
2. Or change `build_network()` to take the `NetworkServiceHandle` as input.  

I feel we should address this separately and keep this change for now

Reply:
> My initial patch to upstream required download_block() to take network handle for same reason, but the upstream team did not like it.

Did you explain to them that it is actually kind of necessary to avoid this ugliness? This feels like design flaw that needs to be addressed sooner rather than later. We're writing a bunch of convoluted code and handling edge cases that should not exist in the first place because of that.

Reply:
I am starting a thread with them separately now, let us address this separately

Reply:
I don't see where, do you mind posting a link? It is best to preserve continuity so that it is possible to follow the discussion. Both for me now and for those who might read this PR in the future.

Reply:
StackExchange is typically for questions, for discussions I usually create a thread on GitHub: https://github.com/paritytech/substrate/discussions